From f7dfa529dc5ac131d3ed9d1b29dff5595d6d40aa Mon Sep 17 00:00:00 2001
From: Yong Zhi <yong.zhi@intel.com>
Date: Tue, 10 Apr 2018 17:16:10 -0500
Subject: [PATCH 037/129] media: intel-ipu4: imx355: Fix v4l2-compliance
 DQEVENT fails

Message for Open Source:
Fix v4l2-compliance fails on imx355 camera sensor subdev
for Control ioctls tests.

Message for Internal:
[Issue/Feature] v4l2-compliance -u /dev/v4l-subdev10

"fail: v4l2-test-subdevs: fmt.colorspace == ~0U"
"fail: v4l2-test-subdevs: checkMBusFrameFmt(node, fmt.format)"
"test Active VIDIOC_SUBDEV_G/S_FMT: FAIL"

"fail: v4l2-test-controls: subscribe event for control"
"     'Camera Controls' failed Inappropriate ioctl for device"
"test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: FAIL"

"fail: v4l2-test-controls: failed to find event for control
       'Exposure'"
"test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: FAIL"

[Root Cause/Changes] Missing implementation.

Change-Id: I9c4a83f7839400b71c46d47f2e39b9518b65a32c
Tracked-On: #JICCCN-87
Signed-off-by: Yong Zhi <yong.zhi@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/i2c/imx355.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/imx355.c b/drivers/media/i2c/imx355.c
index 703ecb5..c2c10e0 100644
--- a/drivers/media/i2c/imx355.c
+++ b/drivers/media/i2c/imx355.c
@@ -7,6 +7,7 @@
 #include <linux/pm_runtime.h>
 #include <media/v4l2-ctrls.h>
 #include <media/v4l2-device.h>
+#include <media/v4l2-event.h>
 
 
 #define IMX355_REG_MODE_SELECT		0x0100
@@ -1280,6 +1281,10 @@ static void imx355_update_pad_format(struct imx355 *imx355,
 	fmt->format.height = mode->height;
 	fmt->format.code = imx355_get_format_code(imx355);
 	fmt->format.field = V4L2_FIELD_NONE;
+	fmt->format.colorspace = V4L2_COLORSPACE_DEFAULT;
+	fmt->format.xfer_func = V4L2_XFER_FUNC_DEFAULT;
+	fmt->format.ycbcr_enc = V4L2_YCBCR_ENC_DEFAULT;
+	fmt->format.quantization = V4L2_QUANTIZATION_DEFAULT;
 }
 
 static int imx355_do_get_pad_format(struct imx355 *imx355,
@@ -1537,6 +1542,11 @@ static int imx355_identify_module(struct imx355 *imx355)
 	return 0;
 }
 
+static const struct v4l2_subdev_core_ops imx355_subdev_core_ops = {
+	.subscribe_event = v4l2_ctrl_subdev_subscribe_event,
+	.unsubscribe_event = v4l2_event_subdev_unsubscribe,
+};
+
 static const struct v4l2_subdev_video_ops imx355_video_ops = {
 	.s_stream = imx355_set_stream,
 };
@@ -1549,6 +1559,7 @@ static int imx355_identify_module(struct imx355 *imx355)
 };
 
 static const struct v4l2_subdev_ops imx355_subdev_ops = {
+	.core = &imx355_subdev_core_ops,
 	.video = &imx355_video_ops,
 	.pad = &imx355_pad_ops,
 };
@@ -1693,7 +1704,8 @@ static int imx355_probe(struct i2c_client *client,
 
 	/* Initialize subdev */
 	imx355->sd.internal_ops = &imx355_internal_ops;
-	imx355->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	imx355->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE |
+				V4L2_SUBDEV_FL_HAS_EVENTS;
 	imx355->sd.entity.ops = &imx355_subdev_entity_ops;
 	imx355->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
 
-- 
1.9.1

